File: Readio/src/main/java/ninegle/Readio/user/config/JwtAuthFilter.java
package ninegle.Readio.user.config;

import java.io.IOException;
import java.util.Date;

import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import io.jsonwebtoken.JwtException;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import ninegle.Readio.global.exception.BusinessException;
import ninegle.Readio.global.exception.domain.ErrorCode;
import ninegle.Readio.user.adapter.UserDetail;
import ninegle.Readio.user.dto.TokenBody;
import ninegle.Readio.user.repository.BlackListRepository;
import ninegle.Readio.user.service.JwtTokenProvider;
import ninegle.Readio.user.service.UserService;

@Slf4j

//필터 체인의 경우 빈으로 자동 등록이 되지 않아서 수동으로 등록한 상태 @RequiredArgsConstructor 불가
//요청이 들어왔을 때 한번만 동작한다
public class JwtAuthFilter extends OncePerRequestFilter {
	//토큰 제공자
	private final JwtTokenProvider jwtTokenProvider;
	private final UserService userService;
	private final BlackListRepository blackListRepository;

	public JwtAuthFilter(JwtTokenProvider jwtTokenProvider, UserService userService,
		BlackListRepository blackListRepository) {
		this.jwtTokenProvider = jwtTokenProvider;
		this.userService = userService;
		this.blackListRepository = blackListRepository;
	}

	@Override
	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response,
		FilterChain filterChain) throws ServletException, IOException {

		//나중에 쓸 일이 있을지?
		// AntPathMatcher pathMatcher = new AntPathMatcher();
		// String uri = request.getRequestURI();
		// String method = request.getMethod();

		String token = resolveToken(request); //헤더에서 토큰값 추출

		// 블랙리스트에 등록된 토큰인지 먼저 검사
		if (blackListRepository.findByInversionAccessToken(token)
			.filter(blacklist -> blacklist.getExpiration().after(new Date()))
			.isPresent()) { // 요청한 access 토큰값이 블랙리스트에 있으며 블랙리스트에 해당 토큰 만료 시간이 유효하면 true가 되어 차단함
			throw new JwtException("이 토큰은 블랙리스트에 등록되어 있으므로 사용할 수 없습니다.");
		}

		if (token != null && jwtTokenProvider.validate(token)) {

			// 토큰에서 사용자 정보를 추출
			TokenBody tokenBody = jwtTokenProvider.parseJwt(token);
			UserDetail userDetail = userService.getDetails(tokenBody.getUserId());

			if (userDetail == null) {
				throw new BusinessException(ErrorCode.USER_NOT_FOUND); //404
			}

			//사용자가 입력한 ID/PW를 UsernamePasswordAuthenticationToken으로 감쌈
			Authentication usernamePasswordAuthenticationToken = new UsernamePasswordAuthenticationToken(userDetail,
				token, userDetail.getAuthorities());
			//SecurityContex는 현재 HTTP 요청에 대한 인증 정보를 저장하는 곳으로 사용자 정보를 spring security가 관리 가능하게 해줌
			SecurityContextHolder.getContext().setAuthentication(usernamePasswordAuthenticationToken);

			//기존 인증으로 있던 validate는 토큰이 위조되지 않았는가?, 서명은 맞는가?, 만료됐는가? 같은 기본적인 무결성 검사만 하므로 인증된 사용자란 보장은 되지 않음
			// Spring Security의 필터 체인은 여전히 "이 요청은 인증된 사용자 인지 물어보기 때문에 spring에게 이 사용자는 인증이 되었음을 알려 필터 체인을 통과시킨다
		}
		filterChain.doFilter(request, response);
	}

	//요청을 받아서 헤더가 있다면 "해더에서" 토큰을 추출하는 용도
	private String resolveToken(HttpServletRequest request) {
		String bearerToken = request.getHeader("Authorization");
		log.debug("Bearer token: {}", bearerToken);
		if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
			return bearerToken.substring(7);
		}
		return null;
	}
}

File: Readio/src/main/java/ninegle/Readio/user/config/JwtAuthFilter.kt
package ninegle.Readio.user.config

import io.jsonwebtoken.JwtException
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.extern.slf4j.Slf4j
import ninegle.Readio.global.exception.BusinessException
import ninegle.Readio.global.exception.domain.ErrorCode
import ninegle.Readio.user.adapter.UserDetail
import ninegle.Readio.user.repository.BlackListRepository
import ninegle.Readio.user.repository.UserRepository
import ninegle.Readio.user.service.JwtTokenProvider
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException
import java.util.*

@Slf4j
@Component
// TossApiClient 있는 패키지
@EnableWebSecurity
class JwtAuthFilter(//토큰 제공자
    private val jwtTokenProvider: JwtTokenProvider,
//    private val userService: UserService,
    private val userRepository: UserRepository,
    private val blackListRepository: BlackListRepository
) : OncePerRequestFilter() {

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain)
    {


        val token = resolveToken(request) //헤더에서 토큰값 추출

        // 블랙리스트에 등록된 토큰인지 먼저 검사
        token?.let {
            val blacklist = blackListRepository.findByInversionAccessToken(it)
            if (blacklist != null && blacklist.expiration.after(Date())) {
                throw JwtException("이 토큰은 블랙리스트에 등록되어 있으므로 사용할 수 없습니다.")
            }
        }


        if (token != null && jwtTokenProvider.validate(token)) {
            // 토큰에서 사용자 정보를 추출

            val tokenBody = jwtTokenProvider.parseJwt(token)
            val userEntity = userRepository.findById(tokenBody.userId) ?: throw BusinessException(ErrorCode.USER_NOT_FOUND) //404


//            //사용자가 입력한 ID/PW를 UsernamePasswordAuthenticationToken으로 감쌈
//            val usernamePasswordAuthenticationToken: Authentication = UsernamePasswordAuthenticationToken(
//                userDetail,
//                token, userDetail.authorities
//            )
            val userDetail = UserDetail(
                id = userEntity.get().id,
                email = userEntity.get().email,
                password = userEntity.get().password,
                role = userEntity.get().role
            )
//            val roleList = listOf(SimpleGrantedAuthority("ROLE_USER"))

            val usernamePasswordAuthenticationToken: Authentication =
                UsernamePasswordAuthenticationToken( userDetail, token, userDetail.authorities)

            //SecurityContex는 현재 HTTP 요청에 대한 인증 정보를 저장하는 곳으로 사용자 정보를 spring security가 관리 가능하게 해줌
            SecurityContextHolder.getContext().authentication = usernamePasswordAuthenticationToken

            //기존 인증으로 있던 validate는 토큰이 위조되지 않았는가?, 서명은 맞는가?, 만료됐는가? 같은 기본적인 무결성 검사만 하므로 인증된 사용자란 보장은 되지 않음
            // Spring Security의 필터 체인은 여전히 "이 요청은 인증된 사용자 인지 물어보기 때문에 spring에게 이 사용자는 인증이 되었음을 알려 필터 체인을 통과시킨다
        }
        filterChain.doFilter(request, response)
    }

    //요청을 받아서 헤더가 있다면 "해더에서" 토큰을 추출하는 용도
    private fun resolveToken(request: HttpServletRequest): String? {
        val bearerToken = request.getHeader("Authorization")
        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
            return bearerToken.substring(7)
        }
        return null
    }
}
